internals/cli/ui/io.go

```diff
@@ -65,50 +67,58 @@ func (o UserIO) Prompts() (Reader, Writer, error) {
 	return o.Input, o.Output, nil
 }
 
-// Reader can read input for a CLI program.
-type Reader interface {
-	io.Reader
-	// ReadPassword reads a line of input from a terminal without local echo.
-	ReadPassword() ([]byte, error)
-	IsPiped() bool
+func (o UserIO) IsStdinPiped() bool {
+	return isPiped(o.Input)
 }
 
-// Readln reads 1 line of input from a io.Reader. The newline character is not included in the response.
-func Readln(r io.Reader) (string, error) {
-	s := bufio.NewScanner(r)
-	s.Scan()
-	err := s.Err()
-	if err != nil {
-		return "", ErrReadInput(err)
-	}
-	return s.Text(), nil
+func (o UserIO) IsStdoutPiped() bool {
+	return isPiped(o.Output)
 }
 
-// Writer can write output for a CLI program.
-type Writer interface {
-	io.Writer
-	IsPiped() bool
+type PasswordReader interface {
+	Read(reader io.Reader) (string, error)
 }
 
-// file implements the Reader and Writer interface.
-type file struct {
-	*os.File
+type passwordReader struct{}
+
+// NewPasswordReader returns a reader that reads a string from the terminal without echoing the user input.
+func NewPasswordReader() PasswordReader {
+	return &passwordReader{}
 }
 
-// ReadPassword reads from a terminal without echoing back the typed input.
-func (f file) ReadPassword() ([]byte, error) {
+// Read reads one line of input from the terminal without echoing the user input.
+func (pr *passwordReader) Read(r io.Reader) (string, error) {
```

Comment:
Do we want to use these outside of the`ui` package? Consider making them private. It's functionality is exposed via `AskPassphrase` and `AskSecret`.

Reply:
I agree, I'll make it unexported.